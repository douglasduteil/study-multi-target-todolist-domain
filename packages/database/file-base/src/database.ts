//

import { inject, postConstruct } from "inversify";
import { Config } from "./config";
import { CONFIG_TOKEN } from "./tokens";

//

export class Database {
  constructor(@inject(CONFIG_TOKEN) private config: Config) {}

  @postConstruct()
  public initialize(): void {
    console.log(`check file ${this.config.path}/${this.config.filename}`);
  }
}
