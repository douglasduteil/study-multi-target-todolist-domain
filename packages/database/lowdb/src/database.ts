//

import { NewTodo, Todo, TodoDataSource } from "@todolist/core";
import low from "lowdb";
import { NotFoundError } from "./errors";
import { LowDbShema } from "./models";

//

export class Database implements TodoDataSource {
  db: low.LowdbSync<LowDbShema>;
  constructor(
    adapter: low.AdapterSync<LowDbShema> | low.AdapterAsync<LowDbShema>
  ) {
    this.db = low(adapter);
    this.db.defaults({ index: 0, todos: [] }).write();
  }
  async getTodos(): Promise<ReadonlyArray<Todo>> {
    return this.db.get("todos").value();
  }

  getTodo(id: string): Todo {
    const todo = this.db
      .get("todos")
      .find({ id })
      .value();

    if (!todo) {
      throw new NotFoundError(`Can't find todo ${id}`);
    }

    return todo;
  }

  async addTodo(todo: NewTodo): Promise<Todo> {
    const index = this.db.get("index").value();
    this.db.set("index", index + 1).write();

    const newTodo: Todo = {
      ...todo,
      completed: false,
      createdAt: new Date(Date.now()),
      id: String(index),
      updatedAt: new Date(Date.now())
    };

    this.db
      .get("todos")
      .push(newTodo)
      .write();

    return newTodo;
  }

  updateTodo(id: string, todo: Partial<Todo>): Todo {
    const actual = this.getTodo(id);
    const completedValues: Partial<Todo> =
      !actual.completed && todo.completed
        ? {
            completed: true, // as the new todo is completed indeed
            completedAt: new Date(Date.now())
          }
        : {
            completedAt: undefined // force undefined to lock the value
          };
    const lockedValues: Partial<Todo> = {
      createdAt: actual.createdAt,
      id: actual.id
    };
    const newTodo: Todo = {
      ...actual,
      ...todo,
      ...completedValues,
      updatedAt: new Date(Date.now()),
      ...lockedValues
    };

    this.db
      .get("todos")
      .find({ id })
      .assign(newTodo)
      .write();

    return newTodo;
  }

  removeTodo(id: string): Todo {
    const actual = this.getTodo(id);

    this.db
      .get("todos")
      .remove({ id })
      .write();

    return actual;
  }
}
