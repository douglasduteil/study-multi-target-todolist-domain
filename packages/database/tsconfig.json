{
  "files": [],
  "references": [{ "path": "level" }, { "path": "lowdb" }]
}
