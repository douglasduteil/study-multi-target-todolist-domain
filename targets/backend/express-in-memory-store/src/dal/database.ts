import { TodoInMemory } from "src/models/todo";

//

export const database = {
  todos: [] as TodoInMemory[]
};
