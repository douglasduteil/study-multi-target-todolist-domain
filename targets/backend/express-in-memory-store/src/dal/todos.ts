//

import uuid from "uuid/v4";

import { Todo } from "@todolist/core";
import { TodoInMemory } from "src/models/todo";
import { database } from "./database";

export async function getTodos(): Promise<sTodoInMemory[]> {
  return database.todos;
}

export async function addTodo(newTodo: Todo): Promise<TodoInMemory> {
  const todo: TodoInMemory = { id: uuid(), ...newTodo };
  database.todos.push(todo);
  return todo;
}

export async function removeTodo(id: string): Promise<TodoInMemory> {
  const removedTodoIndex = database.todos.findIndex(todo => todo.id === id);
  const removedTodo = database.todos[removedTodoIndex];
  database.todos.splice(removedTodoIndex, 1);
  return removedTodo;
}

export async function findTodo(todoId: string): Promise<TodoInMemory> {
  const foundTodo = database.todos.find(todo => todo.id === todoId);
  return foundTodo ? foundTodo : Promise.reject(`todo@${todoId} not found`);
}

export async function updateTodo(
  todoId: string,
  newTodo: Partial<Todo>
): Promise<TodoInMemory> {
  const todoIndex = database.todos.findIndex(todo => todo.id === todoId);
  database.todos[todoIndex] = {
    ...database.todos[todoIndex],
    ...newTodo
  };
  return database.todos[todoIndex];
}
