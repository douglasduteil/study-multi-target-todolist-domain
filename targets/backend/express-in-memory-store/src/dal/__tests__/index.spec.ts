//

//

import uuid from "uuid/v4";

import { database } from "../database";
import { addTodo, findTodo, getTodos, removeTodo, updateTodo } from "../todos";

//

jest.mock("uuid/v4");

//

test("#getTodos should return list of todos", async () => {
  // given
  database.todos = [];

  // when
  const expected = await getTodos();

  // then
  expect(expected).toBe(database.todos);
});

//

test("#addTodo should return list of todos", async () => {
  // given
  database.todos = [];
  (uuid as jest.Mock).mockImplementation(() => "foo-123-bar");

  // when
  const expected = await addTodo({ title: "Hello World", completed: false });

  // then
  expect(expected).toMatchSnapshot();
  expect(database.todos).toMatchSnapshot();
});

//

test("#removeTodo should return list of todos", async () => {
  // given
  database.todos = [
    {
      completed: true,
      id: "foo-123-bar",
      title: "Hello World"
    },
    {
      completed: false,
      id: "bar-123-foo",
      title: "Hello World"
    },
    {
      completed: false,
      id: "123-foo-bar",
      title: "Hello World"
    }
  ];

  // when
  const expected = await removeTodo("bar-123-foo");

  // then
  expect(expected).toMatchSnapshot();
  expect(database.todos).toMatchSnapshot();
});

//

test("#findTodo should return todo@bar-123-foo", async () => {
  // given
  database.todos = [
    {
      completed: true,
      id: "foo-123-bar",
      title: "Hello World"
    },
    {
      completed: false,
      id: "bar-123-foo",
      title: "Hello World"
    },
    {
      completed: false,
      id: "123-foo-bar",
      title: "Hello World"
    }
  ];

  // when
  const expected = await findTodo("bar-123-foo");

  // then
  expect(expected).toMatchSnapshot();
  expect(database.todos).toMatchSnapshot();
});

//

test("#updateTodo should mark as completed todo@bar-123-foo", async () => {
  // given
  database.todos = [
    {
      completed: true,
      id: "foo-123-bar",
      title: "Hello World"
    },
    {
      completed: false,
      id: "bar-123-foo",
      title: "Hello World"
    },
    {
      completed: false,
      id: "123-foo-bar",
      title: "Hello World"
    }
  ];

  // when
  const expected = await updateTodo("bar-123-foo", { completed: true });

  // then
  expect(expected).toMatchSnapshot();
  expect(database.todos).toMatchSnapshot();
});
