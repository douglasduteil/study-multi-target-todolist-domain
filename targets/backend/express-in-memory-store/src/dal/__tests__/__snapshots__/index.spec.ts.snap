// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#addTodo should return list of todos 1`] = `
Object {
  "completed": false,
  "id": "foo-123-bar",
  "title": "Hello World",
}
`;

exports[`#addTodo should return list of todos 2`] = `
Array [
  Object {
    "completed": false,
    "id": "foo-123-bar",
    "title": "Hello World",
  },
]
`;

exports[`#findTodo should return todo@bar-123-foo 1`] = `
Object {
  "completed": false,
  "id": "bar-123-foo",
  "title": "Hello World",
}
`;

exports[`#findTodo should return todo@bar-123-foo 2`] = `
Array [
  Object {
    "completed": true,
    "id": "foo-123-bar",
    "title": "Hello World",
  },
  Object {
    "completed": false,
    "id": "bar-123-foo",
    "title": "Hello World",
  },
  Object {
    "completed": false,
    "id": "123-foo-bar",
    "title": "Hello World",
  },
]
`;

exports[`#removeTodo should return list of todos 1`] = `
Object {
  "completed": false,
  "id": "bar-123-foo",
  "title": "Hello World",
}
`;

exports[`#removeTodo should return list of todos 2`] = `
Array [
  Object {
    "completed": true,
    "id": "foo-123-bar",
    "title": "Hello World",
  },
  Object {
    "completed": false,
    "id": "123-foo-bar",
    "title": "Hello World",
  },
]
`;

exports[`#updateTodo should mark as completed todo@bar-123-foo 1`] = `
Object {
  "completed": true,
  "id": "bar-123-foo",
  "title": "Hello World",
}
`;

exports[`#updateTodo should mark as completed todo@bar-123-foo 2`] = `
Array [
  Object {
    "completed": true,
    "id": "foo-123-bar",
    "title": "Hello World",
  },
  Object {
    "completed": true,
    "id": "bar-123-foo",
    "title": "Hello World",
  },
  Object {
    "completed": false,
    "id": "123-foo-bar",
    "title": "Hello World",
  },
]
`;
